security-admin/src/main/java/org/apache/ranger/security/web/filter/AltiscaleRangerAuthFilter.java

```diff
@@ -0,0 +1,477 @@
+
+		final FilterConfig globalConf = filterConfig;
+		final Map<String, String> params = new HashMap<String, String>();
+		params.put(RANGER_ALGORITHM.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_ALGORITHM));
+		params.put(RANGER_CERTIFICATE_DIR.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_CERTIFICATE_DIR));
+		params.put(RANGER_AUTHENTICATION_PORTAL.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_AUTHENTICATION_PORTAL));
+		params.put(RANGER_ACCOUNT_ID.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_ACCOUNT_ID));
+		params.put(RANGER_CLUSTER_ID.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_CLUSTER_ID));
+		params.put(RANGER_ALT_KERBEROS_ENABLED.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_ALT_KERBEROS_ENABLED));
+		params.put(RANGER_SIGNATURE_SECRET_FILE.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_SIGNATURE_SECRET_FILE));
+		params.put(RANGER_TOKEN_VALIDITY.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_TOKEN_VALIDITY));
+		params.put(RANGER_COOKIE_DOMAIN.replaceFirst(CONFIG_PREFIX+"alt-kerberos.",""), PropertiesUtil.getProperty(RANGER_COOKIE_DOMAIN));
+		params.put(RANGER_NON_BROWSER_USER_AGENTS.replaceFirst(CONFIG_PREFIX,""), PropertiesUtil.getProperty(RANGER_NON_BROWSER_USER_AGENTS));
+		try {
+			params.put("kerberos.principal", SecureClientLogin.getPrincipal(PropertiesUtil.getProperty(PRINCIPAL,""), PropertiesUtil.getProperty(HOST_NAME)));
+		} catch (IOException ignored) {
+			// do nothing
+		}
+		params.put("kerberos.keytab", PropertiesUtil.getProperty(KEYTAB,""));
+		params.put(AUTH_TYPE, "com.altiscale.hadoop.security.AltiscaleAuthenticationHandler");
+		FilterConfig myConf = new FilterConfig() {
+			@Override
+			public ServletContext getServletContext() {
+				if (globalConf != null) {
+					return globalConf.getServletContext();
+				} else {
+					return noContext;
+				}
+			}
+
+			@SuppressWarnings("unchecked")
+			@Override
+			public Enumeration<String> getInitParameterNames() {
+				return new IteratorEnumeration(params.keySet().iterator());
+			}
+
+			@Override
+			public String getInitParameter(String param) {
+				return params.get(param);
+			}
+
+			@Override
+			public String getFilterName() {
+				return "KerberosFilter";
+			}
+		};
+		super.init(myConf);
+		LOG.debug("AltiscaleRangerAuthFilter.init(FilterConfig filterConfig) ==> ended");
+	}
+
+	/**
+	 * This function performs post tasks after {@link AuthenticationFilter#doFilter(ServletRequest, ServletResponse, FilterChain)}
+	 * to create a ranger session if the request is authenticated.
+	 */
+	@Override
+	protected void doFilter(FilterChain filterChain, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+
+		String userName = null;
+		boolean checkCookie = response.containsHeader("Set-Cookie");
+		if(checkCookie){
```

Comment:
refactor 138-168: way too many levels. inner loop should be a short testable method which has a name indicating what it does (self-documentation). Once you refactor it, and write test for it, you'll see how much more readable this method will be.